enum confusion in cvundoes.c

The function FVCopy() in cvundoes.c has its final argument of type "enum fvcopy_type", which has allowed values ct_fullcopy, ct_reference, ct_lookups, and ct_unlinkrefs .  But when this function is called in scripting.c, this argument is frequently given values of type "enum undotype" instead, such as ut_width .  Mixing enums like that is legal in C because they are syntactic sugar over an integer type, and GCC (at least) won't warn about it, but it sure looks like an error.

I don't have a patch to submit - yet - because I haven't traced out everywhere that argument goes.  The FVCopy() function passes the argument into a couple of other functions and it's possible one of them is actually expecting an "enum undotype" value.  There may be some deliberate "cleverness" going on with the undotype values being chosen to be numbers that are not legal values of fvcopy_type.  If so, it's bad style; but at the moment my guess is it's an outright mistake.

The code that passes the wrong-type values seems to have been written by George in 2007.  Submitting this issue so the information will be recorded, and in case anyone else has insights into what's going on.

> Mixing enums like that is legal in C because they are syntactic sugar over an integer type, and GCC (at least) won't warn about it, but it sure looks like an error.

It actually does - just that by default, most warnings aren't enabled when FontForge is compiled. If you look at the Travis log, there's actually thousands of warnings that should actually be resolved. For example:

~~~~
gresedit.c:624:21: warning: case value â€˜5â€™ not in enumerated type â€˜enum res_typeâ€™ [-Wswitch]
     case rt_string: case rt_stringlong:
gresedit.c:2570:24: warning: comparison between â€˜enum res_typeâ€™ and â€˜enum res_type2â€™ [-Wenum-compare]
  else if ( info[i].type==rt_coloralpha )
~~~~

I think clang also throws up some better warnings about this.

--- a/fontforge/scripting.c
+++ b/fontforge/scripting.c
@@ -2547,21 +2547,21 @@ static void bCopyUnlinked(Context *c) {
 }
 
 static void bCopyWidth(Context *c) {
-    FVCopy(c->curfv,ut_width);
+    FVCopyWidth(c->curfv,ut_width);
 }
 
 static void bCopyVWidth(Context *c) {
     if ( c->curfv!=NULL && !c->curfv->sf->hasvmetrics )
 	ScriptError(c,"Vertical metrics not enabled in this font");
-    FVCopy(c->curfv,ut_vwidth);
+    FVCopyWidth(c->curfv,ut_vwidth);
 }
 
 static void bCopyLBearing(Context *c) {
-    FVCopy(c->curfv,ut_lbearing);
+    FVCopyWidth(c->curfv,ut_lbearing);
 }
 
 static void bCopyRBearing(Context *c) {
-    FVCopy(c->curfv,ut_rbearing);
+    FVCopyWidth(c->curfv,ut_rbearing);
 }
 
 static void bCopyAnchors(Context *c) {

